src/libcore/alloc/mod.rs

```diff
@@ -0,0 +1,333 @@
+    /// allocator's size or alignment constraints.
+    ///
+    /// Implementations are encouraged to return `Err` on memory exhaustion rather than panicking or
+    /// aborting, but this is not a strict requirement. (Specifically: it is *legal* to implement
+    /// this trait atop an underlying native allocation library that aborts on memory exhaustion.)
+    ///
+    /// Clients wishing to abort computation in response to an allocation error are encouraged to
+    /// call the [`handle_alloc_error`] function, rather than directly invoking `panic!` or similar.
+    ///
+    /// [`handle_alloc_error`]: ../../alloc/alloc/fn.handle_alloc_error.html
+    fn alloc(&mut self, layout: Layout, init: AllocInit) -> Result<MemoryBlock, AllocErr>;
+
+    /// Deallocates the memory denoted by `memory`.
+    ///
+    /// # Safety
+    ///
+    /// `memory` must be a memory block returned by this allocator.
+    unsafe fn dealloc(&mut self, memory: MemoryBlock);
+
+    /// Attempts to extend the memory block.
+    ///
+    /// The behavior of how the allocator tries to grow the memory is specified by [`placement`].
+    /// The first `memory.size()` bytes are preserved or copied as appropriate from `ptr`, and the
+    /// remaining bytes up to the new `memory.size()` are initialized according to [`init`].
+    ///
+    /// [`placement`]: ReallocPlacement
+    /// [`init`]: AllocInit
+    ///
+    /// # Safety
+    ///
+    /// * `memory` must be a memory block returned by this allocator.
+    // We can't require that `new_size` is strictly greater than `memory.size()` because of ZSTs.
+    // An alternative would be
+    // * `new_size must be strictly greater than `memory.size()` or both are zero
+    /// * `new_size` must be greater than or equal to `memory.size()`
+    /// * `new_size`, when rounded up to the nearest multiple of `memory.align()`, must not overflow
+    ///   (i.e., the rounded value must be less than `usize::MAX`).
+    ///
+    /// [*currently allocated*]: #currently-allocated-memory
+    /// [*fit*]: #memory-fitting
+    ///
+    /// # Errors
+    ///
+    /// Returns `Err` if the new layout does not meet the allocator's size and alignment
+    /// constraints of the allocator, or if growing otherwise fails.
+    ///
+    /// Implementations are encouraged to return `Err` on memory exhaustion rather than panicking or
+    /// aborting, but this is not a strict requirement. (Specifically: it is *legal* to implement
+    /// this trait atop an underlying native allocation library that aborts on memory exhaustion.)
+    ///
+    /// Clients wishing to abort computation in response to an allocation error are encouraged to
+    /// call the [`handle_alloc_error`] function, rather than directly invoking `panic!` or similar.
+    ///
+    /// [`handle_alloc_error`]: ../../alloc/alloc/fn.handle_alloc_error.html
+    unsafe fn grow(
+        &mut self,
+        memory: &mut MemoryBlock,
```

Comment:
If we reduce the size of `MemoryBlock` to 2 words like I suggested above, it will be more efficient to return the new `MemoryBlock` as part of the `Result`.